### PR TITLE
refactor docstring generation code

### PR DESCRIPTION
**Pull request by davidhewitt**
As a first step towards #2866 this is a tweak to the macro code which generates Python docstrings. This PR refactors the behaviour so that instead of always creating a `concat!` expression to generate a nul-terminated string, this will only happen if a Rust 1.54+ macro doc is present (e.g. `#[doc = include_str!(...)]`).

The default case now just collects all the `#[doc]` attributes into a single string.

This should make it easier to factor out the `text_signature` formatting, and avoids wasting compile time invoking the `concat!` macro when not necessary.

### CODE REVIEW

File: pyo3-macros-backend/src/utils.rs [hunk 1]
```diff
@@ -57,7 +57,7 @@ pub fn option_type_argument(ty: &syn::Type) -> Option<&syn::Type> {
     None
 }
 
-/// A syntax tree which evaluates to a null-terminated docstring for Python.
+/// A syntax tree which evaluates to a nul-terminated docstring for Python.
 ///
 /// It's built as a `concat!` evaluation, so it's hard to do anything with this
```
Comment (mejrs) [inline pyo3-macros-backend/src/utils.rs:62]: So these comments are no longer true? 🙂

File: pyo3-macros-backend/src/utils.rs [hunk 2]
```diff
@@ -70,57 +70,71 @@ pub fn get_doc(
...
+        syn::token::Bang(Span::call_site()).to_tokens(&mut tokens);
+        syn::token::Bracket(Span::call_site()).surround(&mut tokens, |tokens| {
+            parts.to_tokens(tokens);
+            syn::token::Comma(Span::call_site()).to_tokens(tokens);
+            syn::LitStr::new("\0", Span::call_site()).to_tokens(tokens);
+        });
+
+        PythonDoc(tokens)
+    } else {
+        // Just a string doc - return directly will nul terminator
```
Comment (mejrs) [inline pyo3-macros-backend/src/utils.rs:134]:
```suggestion
        // Just a string doc - return directly with nul terminator
```

**Review by mejrs — APPROVED** [review]

File: pyo3-macros-backend/src/utils.rs (hunk 1 above)
Comment (davidhewitt) [reply pyo3-macros-backend/src/utils.rs:62]: Good catch, will update.

**Comment by davidhewitt** [comment]
bors r=mejrs

**Pull request merged**